packages/lit-html/src/parts.ts

```diff
@@ -0,0 +1,117 @@
+import {NodePart, nothing} from './lit-html.js';
+
+/**
+ * The state of a NodePart, which can be detached and reattached.
+ */
+export type NodePartState = {};
+
+/**
+ * The private interface for NodePartState, which should be kept opaque.
+ */
+type NodePartStateInternal = {
+  __value: unknown;
+  __fragment: DocumentFragment;
+};
+
+export const detachNodePart = (part: NodePart): NodePartState => {
+  const fragment = document.createDocumentFragment();
+  const state: NodePartStateInternal = {
+    __value: part.__value,
+    __fragment: fragment,
+  };
+  let start = part.__startNode.nextSibling;
+  let nextNode;
+  while (start !== part.__endNode) {
+    nextNode = start!.nextSibling;
+    fragment.append(start!);
+    start = nextNode;
+  }
+  part.__value = nothing;
+  return state;
+};
+
+export const restoreNodePart = (part: NodePart, state: NodePartState) => {
+  // TODO (justinfagnani): make an interal-only interface
+  (part as any).__commitNode((state as NodePartStateInternal).__fragment);
+  part.__value = (state as NodePartStateInternal).__value;
+};
+
+const createMarker = () => document.createComment('');
+
+export const createAndInsertPart = (
+  containerPart: NodePart,
+  refPart?: NodePart
+): NodePart => {
+  const container = containerPart.__startNode.parentNode as Node;
+
+  const endNode =
+    refPart === undefined ? containerPart.__endNode : refPart.__startNode;
+
+  const startNode = container.insertBefore(createMarker(), endNode);
+
+  container.insertBefore(createMarker(), endNode);
+  const newPart = new NodePart(startNode, endNode, containerPart.options);
```

Comment:
Shouldn't `endNode` passed to `NodePart` be the marker created on the line above (not the `ref` node, which may be the next part's start marker)?

The `endNode` on L47 is named wrong, it should be the `refNode` used in this function for insertion, and `endNode` for this part will be the node created on L52.

Reply:
Fixed

Reply:
Sorta curious how this didn't break any tests.  A `clear` would have removed the marker... seems like that should have broken something in `repeat` tests. Would be nice to make sure we have a regression test for this.

Reply:
Me too. I'll try to add a test. Maybe rendering a repeat() then a non-repeat value.

Reply:
Added a test that fails on the bad impl